armory/instrument/config.py

```diff
@@ -29,6 +29,7 @@ def __init__(
         include_adversarial=True,
         include_targeted=True,
         record_metric_per_sample=False,
+        max_record_size=2**20,
```

Comment:
my method of testing this PR was to do the same thing that brought the issue to my attention: add a `hub.record()` in the model's `forward()` that records an image and see whether the output JSON file size balloons.

When I ran this test (using CIFAR), I noticed that the output JSON was still blowing up in size. I realized that this was simply because a `(1, 3, 32, 32)` array is still below the threshold. So for other scenarios with bigger inputs, it'll ignore the arrays (since I imagine they'll exceed `2 ** 20` bytes), but it may be worth considering a stricter threshold. Not sure how arbitrary `2 ** 20` was

Reply:
I put something that was about 4x lower than the `(200, 200, 3)` images. My main concern with pushing too small is making sure that complex records don't get unintentionally removed.

I figured that 1 MB per record would mean that even if you pushed in a bunch of 1 MB records, you'd probably wouldn't get too large of a file (unless you did ~10k records at the 1 MB limit). For CIFAR, the json would still be only ~300 MB if you output the entire test set.

Reply:
If you have a different number that you think works better, happy to add it. But I'm not sure about going so small as needed to catch the CIFAR (let alone MNIST) images. A metric record that's outputting something complex like confusion matrices or a sweep of (tpr, fpr, threshold) values for a ROC Curve could easily go above 3x32x32 = 3072 values.

Reply:
I'll point out that, in the scenario when `hub.record()` is placed inside the model's `forward()`, this gets called multiple times per attack so the file size will grow well beyond 300MB which I've tested/confirmed (hence why it's better to accomplish this user goal with a meter and stage filters, but the point here is that there's nothing stopping a user from using `record()`.

Nonetheless, I agree with your point about being careful not to filter out things that, as you said, could easily exceed 3072 values. Do you think it could make sense to add a constraint on the total size of `self.records` as opposed to just the individual `record`?

Reply:
I'm less confident on what a good value would be here. And it would be rough to go through an entire long scenario run to only figure out at the end that the full record was too large. I would rather have it successfully write the giant thing to disk.

What do you think about a warning that runs at the end that says that the results dictionary is very large - check to ensure large arrays are not being written?

Alternatively, we could keep running track of the size of the records thus far, and have it drop records once it exceeds a certain large size. (But what size would that be? 100 MB? Would it keep running? Error Out?)

Reply:
I'm good with a warning. It does feel a bit arbitrary picking a threshold, so at least the consequences (of a "false positive") we can live with moreso if it's a warning. 100MB seems reasonable

Reply:
I put the warning in `scenario.py`, since that is what is writing results to disk, and set it at 128 MB. Since it is a warning, I did not make it configurable.